RE-35 Snapshot Support

hughsaunders opened the pull request:
This commit adds support to RPCO for saving a snapshot of a build
and reusing it later.

There are two parts:
1) Pre-Snapshot preparation: This is done at the end of a build to
prepare an instance for snapshotting. This is added to the post
hook of the pre_merge_test hook (which is also used for post merge).
RE_JOB_TRIGGER is used to determine the cause of the build, if the
cause is PUSH, then snapshot prep is enabled. Snapshot prep writes
the name of the image to be created to a file, the existence of that
file signals to rpc-gating that an image should be saved.

2) Thaw: These are tasks performed on a new instance created from
a snapshot. The thaw hook is /gating/thaw/run relative to the
instance's rootfs. The thaw hook is implemented in
rpc-openstack/gating/thaw/run. Pre snapshot prep ensures that script
is linked to /gating/thaw/run. The thaw hook fixes known issues
with openstack after moving to a different host, it is a shell
script that uses two playbooks, in the same directory.

Issue: [RE-35](https://rpc-openstack.atlassian.net/browse/RE-35)

hughsaunders commented:
Ignore container_ tests, they have been removed from jenkins.

odyssey4me commented on gating/pre_merge_test/post:
```diff
@@ -55,3 +55,40 @@ if which lxc-ls &> /dev/null; then
+  target="$(readlink -f rpc-openstack)"
+  if [[ -n "$target" ]]; then
+      echo "/opt/rpc-openstack is linked into $WORKSPACE/rpc-openstack which"
+      echo "will be wiped prior to snapshoting."
```
``will be wiped prior to creating a snapshot`` would be a better phrase to avoid the non-word ``snapshotting``

hughsaunders replied:
Done.

mattt416 commented on gating/pre_merge_test/post:
```diff
@@ -55,3 +55,40 @@ if which lxc-ls &> /dev/null; then
+
+  rpc_release="$(extract_rpc_release </opt/rpc-openstack/group_vars/all/release.yml)"
+
+  echo "rpc_${rpc_release}" > /gating/thaw/image_name
```
As mentioned in https://github.com/rcbops/rpc-gating/pull/549/files#r153919388, how do we handle releases that support multiple distros?

hughsaunders replied:
Good point, it would probably be useful to put the distro name in the image name, will update.

hughsaunders replied:
Done.

mattt416 commented on gating/thaw/haproxycheck.yml:
```diff
@@ -0,0 +1,37 @@
+      # Check haproxy backends excluding elasticsearch (Future work required
+      # to get elasticsearch working post thaw).
+      # Haproxy prints the stats itteration so if the line is "1" its ignored
+      # Lines where the first fied is hash are comments and are ignored
```
Nit: typos on itteration, fied, may want to add an explicit TODO on the elasticsearch piece.

hughsaunders replied:
Done.

hughsaunders replied:
Added https://rpc-openstack.atlassian.net/browse/RE-1192 for follow up.

mattt416 commented on gating/pre_merge_test/post:
```diff
@@ -55,3 +55,40 @@ if which lxc-ls &> /dev/null; then
+
+  # /root/.ssh is cleared on thaw. We need to store the keys that were there.
+  mkdir -p /opt/root_ssh_backup
+  cp -vr /root/.ssh /opt/root_ssh_backup
```
Certainly don't think it's critical (especially since this is referenced in multiple places), but I think backing up into /root itself would be preferred.  If anyone else agrees, it may be worth updating, otherwise carry on.  :)

hughsaunders replied:
I didn't know whether the whole of /root is recreated or just /root/.ssh so I moved it out of /root.

hughsaunders commented:
Addressed comments with the following patch: 
```diff
diff --git a/gating/pre_merge_test/post b/gating/pre_merge_test/post
index 14133c87..89b515f5 100755
--- a/gating/pre_merge_test/post
+++ b/gating/pre_merge_test/post
@@ -75,7 +75,7 @@ if [[ "${RE_JOB_TRIGGER:-USER}" == "PUSH" ]]; then
   target="$(readlink -f rpc-openstack)"
   if [[ -n "$target" ]]; then
       echo "/opt/rpc-openstack is linked into $WORKSPACE/rpc-openstack which"
-      echo "will be wiped prior to snapshoting."
+      echo "will be wiped prior to creating a snapshot."
       echo "Removing symlink and copying $WORKSPACE/rpc-openstack to /opt."
       rm rpc-openstack
       cp -ar $target rpc-openstack
@@ -88,7 +88,8 @@ if [[ "${RE_JOB_TRIGGER:-USER}" == "PUSH" ]]; then
   ln -s /opt/rpc-openstack/gating/thaw/run /gating/thaw/run

   rpc_release="$(extract_rpc_release </opt/rpc-openstack/group_vars/all/release.yml)"
+  distro="$(lsb_release --codename --short)"

-  echo "rpc_${rpc_release}" > /gating/thaw/image_name
+  echo "rpc_${rpc_release}_${distro}" > /gating/thaw/image_name
   echo "### END SNAPSHOT PREP ###"
 fi
diff --git a/gating/thaw/haproxycheck.yml b/gating/thaw/haproxycheck.yml
index 8f71da7e..a7e12d1d 100644
--- a/gating/thaw/haproxycheck.yml
+++ b/gating/thaw/haproxycheck.yml
@@ -4,10 +4,10 @@
   - name: Check state of haproxy backends
     shell: |
       set -xeu
-      # Check haproxy backends excluding elasticsearch (Future work required
-      # to get elasticsearch working post thaw).
-      # Haproxy prints the stats itteration so if the line is "1" its ignored
-      # Lines where the first fied is hash are comments and are ignored
+      # Check haproxy backends excluding elasticsearch
+      # TODO: Fix elk in thaw process.
+      # Haproxy prints the stats iteration so if the line is "1" its ignored
+      # Lines where the first field is hash are comments and are ignored
       # Field 6 is srv_op_state, and 2 is fully up.

       # http://cbonte.github.io/haproxy-dconv/1.7/management.html#9.3-show%20servers%20state
```

hughsaunders commented:
Rebased.

hughsaunders commented:
Rebased to pull in lint fixes.

hughsaunders commented:
recheck_xenial_swift_deploy (timed out)

hughsaunders commented:
recheck xenial_swift_deploy

hughsaunders commented:
recheck_xenial_swift_deploy recheck_artifact
